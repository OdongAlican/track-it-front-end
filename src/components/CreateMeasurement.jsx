import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import { createMeasurement } from '../actions/measurementsAction'
import { Link } from 'react-router-dom'

const CreateMeasurement = (props) => {
    const activityId = props.location.state
    const dispatch = useDispatch()
    const [watch, setWatch] = useState('00:00:00:00')
    let [ timer, setTimer ] = useState(false)
    let millisecond = 0
    let second = 0
    let minute = 0
    let hour = 0
    
    const run = () => {
        setWatch( `${ hour < 10 ? "0" + hour : hour }:
                    ${ minute < 10 ? "0" + minute : minute }:
                    ${ second < 10 ? "0" + second : second }:
                    ${ millisecond < 10 ? "0" + millisecond : millisecond }` 
                )
        millisecond++;
        if(millisecond === 100){
            millisecond = 0;
            second++
        }
        if(second === 60 ){
            second = 0
            minute++
        }
        if(minute === 60){
            minute = 0
            hour++
        }
    }

    const startWatch = () => {
        if(!timer){
            setTimer(setInterval(run, 10)) 
        }
    }

    const stopWatch = () => {
        clearInterval(timer)
        setTimer(false)
        
        let duration = (parseInt(watch.toString().split(':')[0].trim()) + 
        ( parseInt(watch.toString().split(':')[1].trim()) / 60) +
        ( parseInt(watch.toString().split(':')[2].trim()) / 3600)).toFixed(2)
        let date = new Date()
        dispatch(createMeasurement({duration, date }, activityId))

    }
    return (
        <div>
            <div className="col-md-4 mx-auto mt-4 border p-3">
            <div className="h1 d-flex justify-content-center">{ watch }</div>
            <div className="d-flex justify-content-center">
                <button onClick={ startWatch } className="btn btn-secondary mr-1">Start</button>
                <button onClick={ stopWatch } className="btn btn-secondary mr-1">
                    <Link to={{
                        pathname: `/activity/${ activityId }/measurements`,
                        state : activityId
                    }}>
                        Stop
                    </Link>
                </button>
            </div>
            </div>
        </div>
    )
}

export default CreateMeasurement