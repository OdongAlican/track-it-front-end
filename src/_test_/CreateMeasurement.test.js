import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import CreateMeasurement from '../components/CreateMeasurement';

afterEach(cleanup);
const initialState = {
  authReducer: {
    output: 10,
    error: 'test',
  },
};

const mockStore = configureStore();
let store;

const location = {
  state: 'sample',
};

it('renders CreateMeasurement correctly', () => {
  store = mockStore(initialState);
  const { getAllByTestId } = render(
    <Router>
      <Provider store={store}><CreateMeasurement location={location} /></Provider>
      ,
    </Router>,
  );
  const result = getAllByTestId('appCreateMeasurement');
  expect(result[0])
    .toHaveTextContent('Measure 00:00:00:00StartStopHomeCreate ActivityInformationMore');
});

it('renders CreateMeasurement correctly', () => {
  store = mockStore(initialState);
  const { getAllByTestId } = render(
    <Router>
      <Provider store={store}><CreateMeasurement location={location} /></Provider>
      ,
    </Router>,
  );
  const result = getAllByTestId('appCreateMeasurement');
  expect(result[0]).not
    .toHaveTextContent('Some Random Text');
});

it('matches CreateMeasurement snapshot', () => {
  const appointment = renderer.create(
    <Router>
      <Provider store={store}><CreateMeasurement location={location} /></Provider>
      ,
    </Router>,
  );
  const tree = appointment.toJSON();
  expect(tree).toMatchSnapshot();
});
