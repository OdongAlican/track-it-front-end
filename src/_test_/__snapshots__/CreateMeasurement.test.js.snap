// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches CreateMeasurement snapshot 1`] = `
Array [
  <div
    className="measure-section"
    data-testid="appCreateMeasurement"
  >
    <div
      className="activity-header"
    >
      <p
        style={
          Object {
            "color": "grey",
          }
        }
      >
        Measure
         
      </p>
    </div>
    <div
      className="mx-auto mt-4 clock-border"
    >
      <div
        className="h1 d-flex justify-content-center"
      >
        <p>
          00:00:00:00
        </p>
      </div>
    </div>
    <div
      className="d-flex justify-content-center mt-4 bg-white p-2 lower-section-button"
    >
      <button
        className="start-watch mr-1 text-white"
        onClick={[Function]}
        type="button"
      >
        Start
      </button>
      <a
        className="text-white stop-watch-link"
        href="/activity/undefined/measurements"
        onClick={[Function]}
      >
        <button
          className="stop-watch mr-1 text-white"
          onClick={[Function]}
          type="button"
        >
          Stop
        </button>
      </a>
    </div>
    <div
      className="footer"
      data-testid="appFooter"
    >
      <a
        aria-current={null}
        className="nav-link"
        href="/activities"
        onClick={[Function]}
        style={
          Object {
            "color": "#ABAEB3",
            "padding": 0,
            "textDecoration": "none",
          }
        }
      >
        <div
          className="link-div"
          data-testid="appButton"
        >
          <i
            className="far fa-chart-bar footer-btn"
          />
          <p>
            Home
          </p>
        </div>
      </a>
      <a
        aria-current={null}
        className="nav-link"
        href="/create-activity"
        onClick={[Function]}
        style={
          Object {
            "color": "#ABAEB3",
            "padding": 0,
            "textDecoration": "none",
          }
        }
      >
        <div
          className="link-div"
          data-testid="appButton"
        >
          <i
            className="fas fa-chart-line footer-btn"
          />
          <p>
            Create Activity
          </p>
        </div>
      </a>
      <a
        aria-current={null}
        className="nav-link"
        href="/information"
        onClick={[Function]}
        style={
          Object {
            "color": "#ABAEB3",
            "padding": 0,
            "textDecoration": "none",
          }
        }
      >
        <div
          className="link-div"
          data-testid="appButton"
        >
          <i
            className="fas fa-chart-pie footer-btn"
          />
          <p>
            Information
          </p>
        </div>
      </a>
      <a
        aria-current={null}
        className="nav-link"
        href="/activities"
        onClick={[Function]}
        style={
          Object {
            "color": "#ABAEB3",
            "padding": 0,
            "textDecoration": "none",
          }
        }
      >
        <div
          className="link-div"
          data-testid="appButton"
        >
          <i
            className="fas fa-ellipsis-h footer-btn"
          />
          <p>
            More
          </p>
        </div>
      </a>
    </div>
  </div>,
  ",",
]
`;
